### ISSUE TITLE
option 'keepalive' should be false in order to create Storage in browser

### ISSUE THREAD

**Issue by joopringelberg** [qgustavor/mega#205]
I found that 
`new Storage({email: accountName, password, userAgent: "Perspectives"}`
will cause endlessly repeating failures because of a 'blocked:mixed-content' error. This is because one has to run the lib in a safe (https) environment, but the 'wait' function that is used to keep the connection alive (I suppose) receives an answer with a url with the 'http' scheme. This is then subsequently called. 
Using 
`new Storage({email: accountName, password, userAgent: "Perspectives", keepalive: false}`
prevents this.
It might be useful to add this hint to the documentation (since, If I am not mistaken, you do not longer work on the source itself).
Kind regards,
Joop Ringelberg

**Comment by qgustavor** [comment]
Transferred from megajs-docs to here, since, while it was first just an issue in the documentation, it's clear that's an implementation issue. I guess the issue is [in the pull method of the API class](https://github.com/qgustavor/mega/blob/76cd6e2b29ec91b903c9490186545ca8d46f8798/lib/api.mjs#L141C32-L144), like with the download and upload methods, there must be some parameter to force the API to return a HTTPS URL.

Sadly MEGA's website code is a mess and, the fact they use abbreviations for everything and barely any comments makes it hard to look into their code. The part that handles `resp.w` in their code is [in the middle of the "crypto" code](https://github.com/meganz/webclient/blob/761b0ea193920f0ede859f5e17e2b957302947ad/js/crypto.js#L505-L508) but I can't find where the equivalent code for [this part](https://github.com/qgustavor/mega/blob/76cd6e2b29ec91b903c9490186545ca8d46f8798/lib/api.mjs#L141C32-L144) in MEGA's code.

**Comment by joopringelberg** [comment]
That is unfortunate. They should support you! Your library is important work.

**Comment by qgustavor** [comment]
MEGA assumes developers should use the official SDK, which is written in C++ so that's not really a fit for browsers. Maybe someone can write a Node.js wrapper for it, but if they themselves are not using this SDK for their web client, is because it's not fit for a browser.

Anyway, I tried to diagnose this using developer tools only to find their web client doesn't call the "sc" endpoint, instead it calls the "wsc" endpoint. Seems the code relevant to it is [here](https://github.com/meganz/webclient/blob/761b0ea193920f0ede859f5e17e2b957302947ad/js/utils/api.js#L1231-L1252).

It called the endpoint with those query arguments: `v=3&ec=&sid=[my sid]&sn=[my sn]`. It returned an object like this:

```json
{
    "a": [
        {
            "a": "ua",
            "st": "b.UHn=",
            "u": "[redated]",
            "ua": [
                "^!stbmp"
            ],
            "v": [
                "[redated]"
            ]
        }
    ],
    "w": "https://g.api.mega.co.nz/wsc/[redated]",
    "sn": "[redated]"
}
```

That's just all I found up to far by checking the developer console. From the arguments in the query string, there isn't something that look like the `ssl` parameter from uploading and downloading. Someone will have to find it in the code. It doesn't help their "crypto.js" file is 3008 lines long, but at least their code is released. There is also their [C++ SDK code](https://github.com/meganz/sdk)...

And checking the SDK code, [they have documented the "sc" endpoint here](https://github.com/meganz/sdk/blob/075bfce1af6c0057bd1a2dcc67adf0bb0e363f4b/doc/source/internals.rst#server-client-requests). "sc" stands for "server-client" and this endpoint accepts a "ssl=1" parameter exactly as the upload and download methods.

So you can just send a pull request adding this parameter somehow. Maybe as a `forceHttps` argument in Storage (defaulting to `true` if the code detects a secure browser environment like `globalThis.isSecureContext`.

**Comment by qgustavor** [comment]
**I want you to vote:** should we simplify the logic for API.handleForceHttps like below?

```javascript
// Current implementation
function handleForceHttps (userOpt) {
  if (userOpt != null) return userOpt
  if (globalThis.Deno) return false
  return process.env.IS_BROWSER_BUILD
}

// Proposed implementation
function handleForceHttps (userOpt) {
  if (userOpt != null) return userOpt
  return globalThis.isSecureContext
}
```

Advantages: it's simpler and the browser/Deno version behaves the same as the Node.js version.

Disadvantage: HTTP websites will use HTTP to connecting with MEGA... which *at the moment* is less safe, unless MEGA found a way around that issue shown in the docs, then we need to implement it soon as it will be an advantage for Deno and Node.js too as those are sending and receiving files from MEGA using only the static file key, which is terrible for privacy. Also `globalThis.isSecureContext` is not supported in really old browsers.

**Comment by qgustavor** [comment]
I tried to check if MEGA servers implement something to improve privacy over HTTP as they use static encryption keys for files. My test consisted of downloading a file I uploaded without encryption and checking if I could guess what was in the file despite it downloading the file using the [2018's CloudRaid thing from MEGA](https://github.com/meganz/webclient/blob/fcb381433d0e3c27b2192d1c7790e62f55002e7b/js/transfers/cloudraid.js#L132-L147).

As described by MEGA: "CloudRAID files, which means files are split into 5 pieces + XOR, across 6 servers." so I guessed this XOR was generated in the client, thus would make it a bit harder to someone eavesdropping the network to know which files are being downloaded if they had access to the file key previously. But, no, I checked the pieces being downloaded from the servers: they were exactly the same as I uploaded it, so, as I uploaded them without any encryption, they were kept without any encryption with clear strings showing what the file was (a video file).

**Comment by xmok** [comment]
> **I want you to vote:** should we simplify the logic for API.handleForceHttps like below?
> 
> ```js
> // Current implementation
> function handleForceHttps (userOpt) {
>   if (userOpt != null) return userOpt
>   if (globalThis.Deno) return false
>   return process.env.IS_BROWSER_BUILD
> }
> 
> // Proposed implementation
> function handleForceHttps (userOpt) {
>   if (userOpt != null) return userOpt
>   return globalThis.isSecureContext
> }
> ```
> 
> Advantages: it's simpler and the browser/Deno version behaves the same as the Node.js version.
> 
> Disadvantage: HTTP websites will use HTTP to connecting with MEGA... which _at the moment_ is less safe, ~unless MEGA found a way around that issue shown in the docs, then we need to implement it soon as it will be an advantage for Deno and Node.js too as those are sending and receiving files from MEGA using only the static file key, which is terrible for privacy~. Also `globalThis.isSecureContext` is not supported in really old browsers.
> 
> **Bonus advantage found [here](https://github.com/qgustavor/mega/pull/209#issuecomment-2333934598):** `w.api.mega.co.nz` servers use outdated TLS settings so Firefox cannot connect to those via HTTPS, just HTTP, so it would allow applications on HTTP to use Storage in Firefox.
> 
> Why? Maybe the HTTP page is hosted in localhost. To be fair, that's bad from MEGA for not updating their TLS settings and that's bad for anyone still using HTTP nowadays, I often develop using HTTPS nowadays since it's really simple to get a certificate from Let's Encrypt and the claimed performance issues from TLS which MEGA claims are largely outdated. IIRC I only use HTTP for single-shot servers (like `python -m http.server` or Node.js equivalents).

I don't use Deno but the issue itself makes sense - not being able to use the lib due to mixed content is an issue Mega itself should address but the workaround you've come up with seems OK to me as those who are already using modern browsers should be unaffected which will hopefully be the majority. Ought to be a cutoff point for the really old browsers that don't support secure context, methinks. So my vote is 🚢 ship it.

**Comment by qgustavor** [comment]
Fixed as of v1.3.1
